File: MicrocontrollerCode/src/common/comms/PacketBroker.cpp
#include "PacketBroker.h"

#include "../telemetry_tasks/SystemStatusTask.h"
#include "../telemetry_tasks/ClockTask.h"
#include "../telemetry_tasks/FilesystemTask.h"
#include "../telemetry_tasks/ImuTask.h"

namespace Cesium {

size_t PacketBroker::routed_packets = 0;

Topic PacketBroker::route_packet(BasePacket &packet)
{
    Topic topic = (Topic)packet.get_topic();
    switch(topic) {
    case Topic::SYSTEM_STATUS:
        DEBUG("Received System Status Packet - ");
        SystemStatusTask::route_packet(packet);
        break;

    case Topic::CLOCK:
        DEBUG("Received Clock Packet - ");
        ClockTask::route_packet(packet);
        break;

    case Topic::FILESYSTEM:
        DEBUG("Received IMU Packet - ");
        FilesystemTask::route_packet(packet);
        break;

    case Topic::IMU:
        DEBUG("Received IMU Packet - ");
        ImuTask::route_packet(packet);
        break;



    // case Topic::TEST_ROCKETS:
    //     DEBUG("Received Test Rocket Packet - ");
    //     TestRocketTask::route_packet(packet);
    //     break;


    default:
        DEBUGLN("Did not route packet");
        SystemStatusTask::send_nack("Topic NotImplemented");       
    }
    

    routed_packets++;

    return topic;
    
}

}

File: PyCommander/telemetry/packet_broker.py
from .command_schema import *
from .base_packet import BasePacket, PacketError
from .packets import *


class Broker:

    

    @staticmethod
    def route_packet(packet: BasePacket) -> Topic:

        topic = Topic(packet.topic)

        match topic:
            case Topic.SYSTEM_STATUS:
                print("Received System Status BasePacket - ", end="")
                SystemStatusTask.route_packet(packet)

            case Topic.POWER:
                print("Received Power BasePacket - ", end="")

            case Topic.RADIO:
                print("Received Radio BasePacket - ", end="")

            case Topic.CLOCK:
                print("Received Clock BasePacket - ", end="")
                ClockTask.route_packet(packet)

            case Topic.NETWORK:
                print("Received Network BasePacket - ", end="")

            case Topic.FILESYSTEM:
                print("Received Filesystem BasePacket - ", end="")

            case Topic.SCHEDULER:
                print("Received Scheduler BasePacket - ", end="")

            case Topic.GPS:
                print("Received GPS BasePacket - ", end="")

            case Topic.IMU:
                print("Received IMU BasePacket - ", end="")

            case Topic.BAROMETER:
                print("Received Barometer BasePacket - ", end="")

            case Topic.SHOCK_ACCEL:
                print("Received Shock Accel BasePacket - ", end="")

            case Topic.GNC_NAVIGATION:
                print("Received GNC Navigation BasePacket - ", end="")

            case Topic.GNC_GUIDANCE:
                print("Received GNC Guidance BasePacket - ", end="")

            case Topic.GNC_CONTROL:
                print("Received GNC Control BasePacket - ", end="")

            case Topic.ACTUATORS:
                print("Received Actuators BasePacket - ", end="")

            case Topic.BAD_TOPIC:
                print("Received Bad Topic BasePacket. How did you get here?!?")

            case _:
                print("Did not route packet")
                return Topic.BAD_TOPIC
            

        return topic



